### DISCUSSION TITLE
Need clarification about the transaction prioritization situation

### DISCUSSION THREAD

**Discussion by facundoolano** [tendermint/tendermint#9772] [Q&A] [answered]
Hi! My team recently started building on tendermint and we've been using v0.34 as, as far as I know, that's the latest stable release (and because that happens to be the latest version supported by the Rust bindings we're using). 

We're looking into prioritizing transactions (reordering them based on how much fee they are paying), and I've seen that:
- a prioritized mempool was [announced for v0.35](https://medium.com/tendermint/tendermint-v0-35-introduces-prioritized-mempool-a-makeover-to-the-peer-to-peer-network-more-61eea6ec572d)
- the mempool was then [set to be deprecated](https://github.com/tendermint/tendermint/issues/9388)
- the [v0.34 version of the proto](https://github.com/tendermint/tendermint/blob/v0.34.24/proto/tendermint/abci/types.proto#L215) already had a priority field available in the CheckTx response, which according to the doc serves the purpose of ordering transactions.

So I'm trying to understand if we should expect that field to work on our current v0.34.22 version, if we should aim to bump to a newer version (which?) where the mempool prioritization is available, or if we should skip to an abci++ solution (is that already usable?)

Thanks!

**Comment by thanethomson** [comment, marked as answer]
The prioritized mempool has effectively been considered a temporary measure to be used until such time that ABCI 1.0's Prepare and Process Proposal calls were available in Tendermint v0.37. This is why it's being deprecated and will be removed in Tendermint v0.38. With PrepareProposal, it's possible to implement an application-side mempool that would allow you to order transactions in whatever way you choose.

With the current design of ABCI 1.0, there is a chance that you will not be able to perform _global_ ordering on transactions, as Tendermint only reaps a certain quantity of transactions from the mempool when constructing a proposal, as per: https://github.com/tendermint/tendermint/blob/abbeb919dffd519ac52c10e2762b45db616f085f/state/execution.go#L116

One additional thing we may consider implementing in future is a flag that, when enabled, will cause the node to reap _all_ transactions from the mempool and send them to the application. This would allow for global ordering by the application of all transactions that were in the mempool at the time that the transactions were reaped.

Tendermint v0.37 with ABCI 1.0 (with Prepare/Process Proposal) is currently available as a release candidate, and the Cosmos SDK is in the process of integrating with it. We are hoping to cut the final v0.37.0 release within the next few weeks.

**Comment by facundoolano** [comment]
Thanks for the reply. What about the prioritized mempool, should I expect it to work in v0.34? (I understand it's temporary but we could use it short-term until we get around to upgrade to v0.37)

**Reply by thanethomson** [reply]
It should work in v0.34. If you encounter any problems with it, feel free to open an issue.

**Reply by facundoolano** [reply]
cool, thanks
